drivers/openstack/openstack.go

```diff
@@ -753,7 +778,14 @@ func (d *Driver) assignFloatingIP() error {
 
 	var floatingIP *FloatingIP
 
-	log.Debugf("Looking for an available floating IP", map[string]string{
+	// This is currently broken
+	// When using parallel docker-machine create calls, it might get the same IP
+	// and fail to update the port, but somehow Update() does not fail
+	// and it will think the same IP has been given to different VMs
+	// https://github.com/docker/machine/issues/4038
+
+	_ = ips
+	/*log.Debugf("Looking for an available floating IP", map[string]string{
```

Comment:
We shouldn't have dead code ; either 
1. This is never useful and it should be removed
2. It does something useful but breaks in some cases, in which case it needs to be rewritten and/or fixed.

Reply:
I'd really like to have this fixed, too. 
I've tried digging around to see what the actual problem is.
I've got so far:
assignFloatingIP() from openstack.go calls d.client.GetFloatingIPs().
This actually gets associated FloatingIPs if there are some.
This result is given to to client.go via d.client.AssignFloatingIP()
AssignFloatingIP() will then calls assignNeutronFloatingIP().
assignNeutronFloatingIP then either:
 - allocates  new IP if none were given as parameter by calling floatingips.Create()
   - this will actually fill up all floating IP quota because deleteFloatingIP() was not implemented yet
 - updates the floatingIP by calling floatingips.Update() and passing the portID of the VM that should get the allocated IP

So now floatingips.Update() will succeed if it is the only occurrence at this time, but when running this in parallel (this is what gitlab-runner does when spawning multiple VMs at the same time) floatingIPs.Update() will *always* succeed wether or not the floating IP could be updated with the port or not.

This is where I have no idea how to fix this. https://github.com/gophercloud/gophercloud/blob/master/openstack/networking/v2/extensions/layer3/floatingips/requests.go#L132 actually looks valid enough that it should fail if it could not update the port.

Do you have any hints for this?

Reply:
I see. Essentially we have N processes requesting a list of IPs from a remote resource and each claiming the first resource from the list as their own. One solution may be to create a lock file of some kind that lists which IPs have been claimed, but even that wouldn't be foolproof (if the concurrent requests happen in different VMs/containers for example).

The crux of the issue IMO is that `docker-machine` was never really designed to issue parallel `create` commands.

Reply:
Actually the best solution might be to lock the whole `assignFloatingIP` sequence behind a lockfile using something like https://godoc.org/github.com/nightlyone/lockfile

That'll probably slow down the parallel execution scenario but slow's better than broken.

Reply:
Couldn't it also be a solution to always request a new floating ip, and fully releasing the floating ip after use? That would be much simpler and allow parallel operation even with multiple VMs/ containers.

Reply:
That is exactly what I have implemented in this pull request, but I left the old and currently broken (when used in parallel) logic commented out that you see there.

```
As a workaround we can always allocate new floating IPs and delete them
properly when removing the VM.
``` :-)

If we can agree on this as only solution I could just remove the old logic.

Reply:
If the old logic is necessary for some users, you could also add some way to configure the floating ip strategy or something like that. But I can't think of a really good use case.

Reply:
I'm running into this issue using gitlab-runner autoscaling docker-machine executor against multiple OpenStack private clouds. We pre-allocate enough floating IPs for the max number of concurrent builds we run so we have a known set of IPs for CI build network traffic. Maybe that's not a "really good" use case but it is what it is. :)

+1 for a lockfile

Reply:
that use case makes sense! 👍

Reply:
@stefanandres I think it's fine to just delete the old logic here. Commented out code isn't used anyway :)